### PR TITLE
repo-updater: Improved repository handling for Github external services

### PR DESCRIPTION
**Pull request by tsenart**
This change-set is a test-bed for using the new repository syncing code with a single external service kind, namely, Github.

Once we nail this one together, @keegancsmith and I will split the remainder of the code hosts between ourselves.

Part of #2025

#### TODO **before** merging this PR:

- [x] Add feature flag.
- [x] Add and test migrations to drop "global_dep" and "pkgs" tables. These migrations need to come before the migration that adds a unique (external_service_type, external_service_id, external_id) constraint to the repo table because it only knows how to update foreign key constraints on the "discussion_threads_target_repo" table and it'd fail because of the other existing foreign key constraints on "global_dep_repo_id" and "pkgs_repo_id".
- [x] Make `api.ExternalServicesListRequest{Kind -> Kinds}` change backwards compatible. (Keep old field `Kind` for the roll-out)
- [x] Address TODO around repos being renamed that are marked as deleted.
- [x] Test feature flag.
- [x] Ensure that repos are always upserted with `external_service_xxx` columns fully set in the old code.
- [x] Review the PR

#### Deployment notes

- Deployment on sourcegraph.com will likely require some hand-holding because the migrations will take relatively long. 
- Ensure that Kubernetes timeouts don't kill the migrations mid-way.

#### Follow-up work **after** merging this PR:

- [ ] Remove old scheduler https://github.com/sourcegraph/sourcegraph/issues/2063
- [ ] Ensure that sourcegraph.com code paths work. How does manually adding a repo when directly visiting one interact with the new Syncer?
    - What if we always had a github.com external service that can't be deleted, and we extended an external service's configuration with an array "sync" field that could have its values ["poll",  "lazy"]?
    - Would this lift the special case of github.com into better designed model? (e.g `sync: ["lazy"]`)
- [ ] Test: Github Source (go-vcr) and contract tests.
- [ ] Test: Default github.com source added to list of sources yielded from `ExternalServicesSourcer`. How does this interact with the rest of the system? Can we instead always have an external service for github.com that cannot be deleted?
- [ ] Test: Server HTTP endpoints
- [ ] Test: Update from diff in scheduler
- [ ] Implement migrations for repos that would not be yielded by any current external service configuration (source) so that they are not simply removed.
   - Find an external service that has the same URL as the repo's external_service_id, if set, and the domain in the name if not. Add repo to this external service's explicit `github.repos` list.
  - In case the repo's name and external_service_id don't match any existing external service config, create one of the kind Other with these orphaned repos. Derive clone URL from name || external_service_id
- [ ] Remove enable/disable feature from new syncer managed repos.
- [ ] Remove explicit deletion feature for new syncer managed repos.

### CODE REVIEW

File: cmd/repo-updater/repos/diff.go [hunk 1]
```diff
@@ -0,0 +1,65 @@
...
+	Unmodified []Diffable
+}
+
+// A Diffable can be diffed by the NewDiff function.
+type Diffable interface {
+	ID() string
+}
+
+// 1 name: gh/gorilla/mux         external: MUX
+// 2 name: github.com/gorilla/mux external: MUX
```
Comment (tsenart) [inline cmd/repo-updater/repos/diff.go]: @keegancsmith: Just noticed these left-over comments. Please delete tomorrow.

Comment (tsenart) [reply cmd/repo-updater/repos/diff.go]: On the other hand, we still need to find a solution for multiple names of a single repo.

File: cmd/frontend/db/repos.go [hunk 2]
```diff
@@ -311,7 +311,10 @@ func parsePattern(p string) ([]*sqlf.Query, error) {
 }
 
 func (*repos) listSQL(opt ReposListOptions) (conds []*sqlf.Query, err error) {
-	conds = []*sqlf.Query{sqlf.Sprintf("TRUE")}
+	conds = []*sqlf.Query{
+		sqlf.Sprintf("TRUE"),
```
Comment (nicksnyder) [inline cmd/frontend/db/repos.go]: No longer necessary?
```suggestion
```

File: migrations/1528395567_add_repo_external_service_unique_constraint.up.sql [hunk 3]
```diff
@@ -0,0 +1 @@
+ALTER TABLE repo ADD CONSTRAINT repo_external_service_unique UNIQUE (external_id, external_service_type, external_service_id);
```
Comment (nicksnyder) [inline migrations/1528395567_add_repo_external_service_unique_constraint.up.sql]: It would be helpful for me to understand the desired final state of the `repo` table 2 months from now (e.g. which fields will be deprecated/removed, which fields are maintained/added, and what is the meaning of each field).

These columns pre-date the formal notion of external services, so it is unclear to me if you are planning to repurpose these fields or not.

Comment (tsenart) [reply migrations/1528395567_add_repo_external_service_unique_constraint.up.sql]: We don't know yet what it should be in 2 months. It's developing. See https://github.com/sourcegraph/sourcegraph/issues/2025#issuecomment-463228713

Another thing we want to try and validate is to store the full JSON external service metadata of a repo, when available, in a JSONB column. We'd have then conditional indexes into this column depending on the external_service_type that would support filter queries. To be seen in it's doable / worth it / etc.

Comment (nicksnyder) [reply migrations/1528395567_add_repo_external_service_unique_constraint.up.sql]: > We don't know yet what it should be in 2 months.

I don't understand how you would plan to merge this PR before having a plan for what these columns mean moving forward.

File: cmd/repo-updater/repos/source.go [hunk 4]
```diff
@@ -4,120 +4,137 @@ import (
...
-
-// A GithubSource yields repositories from multiple Github connections configured
-// in Sourcegraph via the external services configuration and retrieved
-// from the Frontend API.
-type GithubSource struct {
+// Sourcerer converts each code host connection configured via external services
+// in the frontend API to a Source that yields Repos. Each invocation of ListSources
+// may yield different Sources depending on what the user configured at a given point
+// in time.
+type Sourcerer struct {
```
Comment (nicksnyder) [inline cmd/repo-updater/repos/source.go]: The name of this type is not intuitive or descriptive to me. I can't think of a good name myself so my next question would be: "is this type even necessary"?

It seems like this type could be removed entirely and replaced with a single method:

```
func ListSources(ctx context.Context, api InternalAPI, kinds ...string) ([]Source, error)
```

Or alternatively, attach ListSources directly to Syncer (since that is the only location where this is used).
```
func (s Syncer) listSources(ctx context.Context, kinds ...string) ([]Source, error) {
```

File: migrations/1528395567_add_repo_external_service_unique_constraint.up.sql (hunk 3 above)
Comment (tsenart) [reply migrations/1528395567_add_repo_external_service_unique_constraint.up.sql]: > I don't understand how you would be able to merge this PR before having a plan for what these columns mean moving forward.

Before merging this PR this needs to be sorted out.

File: cmd/repo-updater/repos/source.go (hunk 4 above)
Comment (tsenart) [reply cmd/repo-updater/repos/source.go]: Where's the fun in that?

Sourcerer 

But OK, we'll abstain from cute names :-) Would `Sourcer` be more clear to you? One who sources?

**Comment by nicksnyder** [comment]
I am unsubscribing from notifications for the PR for now. You can re-request review from me (if you want) when it is ready.

**Comment by keegancsmith** [comment]
Thinking through how we migrate repositories when we turn on the syncer. I
think this is the last blocker before merging. We need to implement the logic
which updates the external services configuration / does the migration. The
below is for Github repositories only. Repositories not managed by Syncer yet
are excluded from the migration.

- enabled is the value of the enabled column
- external is if the external columns are set
- source is if there is an external service which includes the repository.
- visible is if the repository should be on Sourcegraph after migration.

| enabled | external | source | visible | migration                                                                                  |
|:-------:|:--------:|:------:|:-------:|--------------------------------------------------------------------------------------------|
| T       | T        | T      | T       | none                                                                                       |
| F       | T        | T      | F       | add to blacklist                                                                           |
| F       | F        | T      | F       | add to blacklist                                                                           |
| T       | F        | T      | T       | on first sync will have external set. Is there an intermediate state where it is disabled? |
| T       | T        | F      | T       | add to whitelist in an external service                                                    |
| F       | T        | F      | F       | none                                                                                       |
| T       | F        | F      | T       | add to whitelist in an external service                                                    |
| F       | F        | F      | F       | none                                                                                       |

Note: The goal is to preserve the enabled state (enabled == visible).

File: cmd/repo-updater/repos/db.go [hunk 5]
```diff
@@ -0,0 +1,191 @@
...
+			return verr
+		}
+		if dirty { // this shouldn't happen, but checking anyways
+			return err
+		}
+		log15.Warn("WARNING: Detected an old version of Sourcegraph. The database has migrated to a newer version. If you have applied a rollback, this is expected and you can ignore this warning. If not, please contact support@sourcegraph.com for further assistance.", "db_version", version)
+		return nil
+	}
+	return err
+}
```
[the remark this answers is not in the document]
Comment (tsenart) [reply cmd/repo-updater/repos/db.go]: This function is only used for creating test databases in https://github.com/sourcegraph/sourcegraph/pull/2266/files/b5b28db343d4055917b8dd41257182f0d323f46f#diff-26749d3dfaa5fcee9ef8be3173766a64R26

File: cmd/repo-updater/repoupdater/server.go [hunk 6]
```diff
@@ -101,10 +105,16 @@ func (s *Server) handleExternalServiceSync(w http.ResponseWriter, r *http.Reques
 	}
 
 	switch req.ExternalService.Kind {
+	case "GITHUB":
+		_, err := s.Syncer.Sync(r.Context())
```
[the remark this answers is not in the document]
Comment (tsenart) [reply cmd/repo-updater/repoupdater/server.go]: Good catch.

File: cmd/repo-updater/repos/scheduler.go [hunk 7]
```diff
@@ -179,6 +179,50 @@ var configuredLimiter = func() *mutablelimiter.Limiter {
 	return limiter
 }
 
+// UpdateFromDiff updates the list of configured repos from the given sync diff.
```
Comment (keegancsmith) [inline cmd/repo-updater/repos/scheduler.go]: @keegancsmith continue reviewing from here

File: cmd/repo-updater/repos/scheduler.go [hunk 8]
```diff
@@ -238,6 +238,50 @@ var configuredLimiter = func() *mutablelimiter.Limiter {
...
+			s.updateQueue.update(repo)
+		} else {
+			log15.Debug("scheduler.update-from-diff.modified.deleted", "repo", mod.Name)
+			s.schedule.remove(repo)
+			updating := false
+			s.updateQueue.remove(repo, updating)
+		}
+	}
+
+	// schedKnownRepos.Set(float64(len(diff.Unmodified) + len(diff.Modified) + len(diff.Added)))
```
Comment (keegancsmith) [inline cmd/repo-updater/repos/scheduler.go]: why did you remove setting this gauge?

File: cmd/repo-updater/repos/scheduler.go [hunk 9]
```diff
@@ -238,6 +238,50 @@ var configuredLimiter = func() *mutablelimiter.Limiter {
...
+		repo := configuredRepo2FromRepo(add)
+		s.schedule.add(repo)
+		s.updateQueue.enqueue(repo, priorityLow)
+	}
+
+	for _, mod := range diff.Modified {
+		if repo := configuredRepo2FromRepo(mod); repo.Enabled {
+			log15.Debug("scheduler.update-from-diff.modified.update", "repo", mod.Name)
+			s.schedule.update(repo)
+			s.updateQueue.update(repo)
```
Comment (keegancsmith) [inline cmd/repo-updater/repos/scheduler.go]: we should only care about the enabled case

File: cmd/repo-updater/repos/scheduler.go [hunk 10]
```diff
@@ -238,6 +238,50 @@ var configuredLimiter = func() *mutablelimiter.Limiter {
 	return limiter
 }
 
+// UpdateFromDiff updates the list of configured repos from the given sync diff.
+func (s *updateScheduler) UpdateFromDiff(diff Diff) {
```
Comment (keegancsmith) [inline cmd/repo-updater/repos/scheduler.go]: I worry that the scheduler is keyed by repo name, but our diff is not. Maybe we should be using the same method as before where instead of using the diff, we create a diff between the schedulers repos and the repos that should be synced?

Additionally what happens to the updateScheduler when both a github and a gitlab source are running? I hope they don't trample on each other.

File: cmd/repo-updater/repos/source.go [hunk 11]
```diff
@@ -0,0 +1,162 @@
...
+			srcs = append(srcs, src)
+		}
+	}
+
+	if !includesGitHubDotComSource(srcs) {
+		// add a GitHub.com source by default, to support navigating to URL paths like
+		// /github.com/foo/bar to auto-add that repository.
+		src, err := NewGithubDotComSource()
+		srcs, errs = append(srcs, src), multierror.Append(errs, err)
+	}
```
Comment (keegancsmith) [inline cmd/repo-updater/repos/source.go]: do we need to do this? do we need NewGithubDotComSource? The github.com source doesn't list  any repositories. It is only used in repoLookup. So doesn't need to be part of the new syncer.

File: cmd/repo-updater/repos/syncer.go [hunk 12]
```diff
@@ -0,0 +1,249 @@
...
+		upserts = append(upserts, repo)
+	}
+
+	for _, repo := range diff.Modified {
+		repo.UpdatedAt, repo.DeletedAt = now, time.Time{}
+		upserts = append(upserts, repo)
+	}
+
+	for _, repo := range diff.Added {
+		repo.CreatedAt, repo.UpdatedAt, repo.DeletedAt = now, now, time.Time{}
```
Comment (keegancsmith) [inline cmd/repo-updater/repos/syncer.go]: could we have a race condition here? We send diff down a channel in Run. Assuming another goroutine is reading that, and we are mutating here ==> race.

File: migrations/1528395570_.down.sql [hunk 13]
```diff
@@ -0,0 +1,8 @@
+ALTER TABLE repo ADD CONSTRAINT check_external CHECK (
+	external_id IS NULL
+  AND external_service_type IS NULL
+  AND external_service_id IS NULL
+  OR external_id IS NOT NULL
+  AND external_service_type IS NOT NULL
+  AND external_service_id IS NOT NULL
```
Comment (keegancsmith) [inline migrations/1528395570_.down.sql]: do you not need some sort of paranthesis here?

File: cmd/repo-updater/repos/scheduler.go (hunk 8 above)
Comment (tsenart) [reply cmd/repo-updater/repos/scheduler.go]: This gauge description is `The number of unique repositories that have been managed by the scheduler.`. We'd have to keep a set for this. Do you think it's worth it?

File: cmd/repo-updater/repos/scheduler.go (hunk 9 above)
Comment (tsenart) [reply cmd/repo-updater/repos/scheduler.go]: For context, this was adapted from `updateSource` originally. I'll remove the else case.

File: cmd/repo-updater/repos/scheduler.go (hunk 10 above)
Comment (tsenart) [reply cmd/repo-updater/repos/scheduler.go]: > I worry that the scheduler is keyed by repo name, but our diff is not.

Is your worry that we won't be able to handle renamed repos in the scheduler? Apart from that, what other risks are pondering?

As for renames, not deduplicating here is the desired behaviour I think (at least for now), since git-server doesn't have dedup logic yet, and so will need to re-clone the repo if its name changed.

> Maybe we should be using the same method as before where instead of using the diff, we create a diff between the schedulers repos and the repos that should be synced?

What benefits do you have in mind?

> Additionally what happens to the updateScheduler when both a github and a gitlab source are running? I hope they don't trample on each other.

What changed here that makes you worry about this?

File: cmd/repo-updater/repos/source.go (hunk 11 above)
Comment (tsenart) [reply cmd/repo-updater/repos/source.go]: Once we remove the feature flag, how do you plan to support the same `repoLookup` functionality?

File: cmd/repo-updater/repos/syncer.go (hunk 12 above)
Comment (tsenart) [reply cmd/repo-updater/repos/syncer.go]: We send the diff down the channel after we call upserts only. Once the diff is sent to the channel, it should be treated as immutable by the owner (passing ownership to the receiver).

Comment (tsenart) [reply cmd/repo-updater/repos/syncer.go]: In fact, it should be treated as immutable by receivers too (read-only).

File: migrations/1528395570_.down.sql (hunk 13 above)
Comment (tsenart) [reply migrations/1528395570_.down.sql]: Good question. I copied what the check constraint was before. I just checked in `psql` and , surprisingly, it seems OK.

```
localhost sourcegraph@sourcegraph=# SELECT false AND false OR true AND true;
 ?column? 
----------
 t
(1 row)

Time: 0.396 ms
```

Looking at https://www.postgresql.org/docs/9.6/sql-syntax-lexical.html#SQL-PRECEDENCE-TABLE, it's confirmed that AND takes precedence over OR.

File: cmd/repo-updater/repos/scheduler.go (hunk 8 above)
Comment (keegancsmith) [reply cmd/repo-updater/repos/scheduler.go]: lets just add a new gauge?
